Upgrade to C++17 [CMake]

tigerw opened the pull request:
+ Enable CMake-native pre-compiled headers
+ Enable unity builds

tigerw commented:
The C++17 changes need to be merged at the same time.

peterbell10 commented on CMake/SetCompilerFlags.cmake:
```diff
@@ -0,0 +1,36 @@
+if(MSVC)
```
What is this file and how is it different from `SetFlags.cmake`?

tigerw replied:
> "Who are you and what have you done with the real SetFlags?" peterbell angrily demands of SetCompilerFlags.

It's a migration of SetFlags that removes functionality now handled natively by CMake and to use targets so we don't have to patch around compiler warnings generated by library code. The original SetFlags was only renamed to avoid polluting the diff; after this is merged I'm gonna do a second round finishing everything up.

peterbell10 replied:
So SetFlags isn't actually used any more? We should keep the same warning level, not just remove the flags.

peterbell10 replied:
You also seem to missing link-time optimisation flags on MSVC.

tigerw replied:
Agreed, right now Wall/extra is enabled but warnings-as-errors isn't yet, I am intending to fix it in a follow up PR alongside PCH (target_precompile_headers) on Tools/.

I can confirm that Whole Program Optimisation on Visual Studio was enabled by setting CMAKE_INTERPROCEDURAL_OPTIMISATION.

tigerw commented on tests/ChunkData/CMakeLists.txt:
```diff
@@ -43,5 +38,5 @@ set_target_properties(
 set_target_properties(
 	ChunkBuffer
-	PROPERTIES FOLDER Lib
+	PROPERTIES FOLDER Tests/ChunkData
```
It's for visuals anyway, sure.

tigerw commented on CMake/GenerateBindings.cmake:
```diff
@@ -0,0 +1,112 @@
+)
+
+# Make the file paths absolute and pointing to the bindings folder:
+set(BINDINGS_FOLDER "${CMAKE_CURRENT_SOURCE_DIR}/src/Bindings/")
```
Thanks

tigerw commented on CMake/AddDependencies.cmake:
```diff
@@ -0,0 +1,123 @@
+add_subdirectory(lib/zlib)
+
+# Add proper includes for LibEvent's event-config.h header:
+target_include_directories(
+	${CMAKE_PROJECT_NAME} SYSTEM
+	PRIVATE lib/fmt/include
+	PRIVATE lib/jsoncpp/include
+	PRIVATE ${LIBEVENT_INCLUDE_DIRS}
+	PRIVATE lib/lua/src
+	PRIVATE lib/mbedtls/include
+	PRIVATE lib/SQLiteCpp/include
+	PRIVATE lib/TCLAP/include
+	PRIVATE lib # TODO fix files including zlib/x instead of x
+)
```
Investigating

tigerw commented:
Summary of changes:

1. Unify sources under the Cuberite target
2. As a result pull Bindings generation up a level
3. As a result pull IDE visual source tree generation up a level
4. Delete code handling generator-specific version setting that is now native to CMake
5. As a result start moving away from SetFlags

peterbell10 commented on src/CMakeLists.txt:
```diff
@@ -169,164 +157,45 @@ SET (HDRS
-if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
+# Link process information library:
+if (WIN32)
+	target_link_libraries(${CMAKE_PROJECT_NAME} PRIVATE Psapi.lib)
```
What happened to `ws2_32.lib`? Not needed?

tigerw replied:
Linked by libevent, I think.

peterbell10 commented on src/CMakeLists.txt:
```diff
@@ -169,164 +157,45 @@ SET (HDRS
-		file(APPEND ${CMAKE_CURRENT_SOURCE_DIR}/Bindings/BindingDependencies.txt "${dependency}\n")
-	endforeach()
-
-	set_directory_properties(PROPERTIES ADDITIONAL_MAKE_CLEAN_FILES "Bindings.cpp Bindings.h")
```
This seems like a nice idea, shame to drop it. Or does CMake clean generated output files automatically now?

tigerw replied:
Yep :D Docs say the makefile generators automatically clean up and generated files, and can confirm during testing.

peterbell10 commented on src/CMakeLists.txt:
```diff
@@ -1,18 +1,8 @@
+target_sources(
+	${CMAKE_PROJECT_NAME} PRIVATE
 
+	Resources/Cuberite.rc
```
This is a windows only thing, right?
```suggestion
	$<$<BOOL:WIN32>:Resources/Cuberite.rc>
```

tigerw replied:
Cygwin/MSYS can process them too, and since it's handled automatically by CMake it can be left in.

peterbell10 commented on src/CMakeLists.txt:
```diff
@@ -169,164 +157,45 @@ SET (HDRS
-	)
-	list(APPEND SOURCE "Resources/Cuberite.rc")
-
-	# Make MSVC generate the PDB files even for the release build:
-	set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} /Zi")
-	set(CMAKE_C_FLAGS_RELEASE   "${CMAKE_C_FLAGS_RELEASE}   /Zi")
-	set(CMAKE_EXE_LINKER_FLAGS_RELEASE    "${CMAKE_EXE_LINKER_FLAGS_RELEASE}    /DEBUG")
-	set(CMAKE_SHARED_LINKER_FLAGS_RELEASE "${CMAKE_SHARED_LINKER_FLAGS_RELEASE} /DEBUG")
-	set(CMAKE_MODULE_LINKER_FLAGS_RELEASE "${CMAKE_MODULE_LINKER_FLAGS_RELEASE} /DEBUG")
```
Should we be building RELEASE_WITH_DEBUG_INFO, or something? Do the windows stack traces still work?

tigerw replied:
dun

peterbell10 commented on CMake/GroupSources.cmake:
```diff
@@ -0,0 +1,42 @@
+
+# Put all files into one project, separate by the folders:
+get_property(TARGET_SOURCE_FILES TARGET ${CMAKE_PROJECT_NAME} PROPERTY SOURCES)
+source_group(TREE ${CMAKE_CURRENT_SOURCE_DIR}/src FILES ${TARGET_SOURCE_FILES})
```
This will probably break with spaces in the path
```suggestion
source_group(TREE "${CMAKE_CURRENT_SOURCE_DIR}/src" FILES ${TARGET_SOURCE_FILES})
```

peterbell10 replied:
Also, please don't use `CMAKE_CURRENT_SOURCE_DIR` when you aren't in a `CMakeLists.txt` file. That's just confusing. I think `Cuberite_SOURCE_DIR` or `PROJECT_SOURCE_DIR` are less ambiguous.

tigerw replied:
Noted, thanks

peterbell10 commented on CMakeLists.txt:
```diff
@@ -251,84 +196,66 @@ if (NOT EXISTS ${PROJECT_SOURCE_DIR}/lib/zlib/CMakeLists.txt)
-# Prettify jsoncpp_lib name in VS solution explorer
+# Prettify jsoncpp_lib name in VS solution explorer:
 set_property(TARGET jsoncpp_lib PROPERTY PROJECT_LABEL "jsoncpp")
-# jsoncpp uses these for ccache support, clashing with our method
-set_property(GLOBAL PROPERTY RULE_LAUNCH_COMPILE "")
-set_property(GLOBAL PROPERTY RULE_LAUNCH_LINK "")
```
Oh, you moved the submodules to master again. Please stick to releases... They exist for a reason.

tigerw replied:
eh, it's not that important for a minecraft server. Jsoncpp recently had a 0.11 release, albeit on another branch, so clearly they consider it stable enough.

peterbell10 replied:
We're on version 1.9.2, so 00.11 is not just a different branch but a completely different major version.

peterbell10 replied:
Oh, actually it looks like 00.11 is branched more recently than 1.9.2 ...
That doesn't make any sense to me, but okay.

peterbell10 replied:
We should still be on a release tag though. Things like `libevent` and `mbedtls` especially.

tigerw replied:
ok, i'll move libevent and embedtls

peterbell10 commented on src/CMakeLists.txt:
```diff
@@ -169,164 +157,45 @@ SET (HDRS
-foreach(arg ${ALLFILES})
-	set(ALLFILESLINES "${ALLFILESLINES}${arg}\n")
-endforeach()
-FILE(WRITE "AllFiles.lst" "${ALLFILESLINES}")
```
This is needed for `CheckBasicStyle.lua`

tigerw replied:
dun

tigerw commented on CMakeLists.txt:
```diff
@@ -13,19 +13,17 @@
-if (MSVC)
-	set_property(GLOBAL PROPERTY USE_FOLDERS ON)
-endif()
+cmake_minimum_required (VERSION 3.13)
```
CMake 3.13 introduced CMP0076, target_sources automatically converts relative to absolute, which is needed behaviour here. We agreed to target the version in Debian's repository as a minimum.

peterbell10 replied:
Then why did you agree to merge it as 3.12.4 in the first place? And what was wrong with the `SRCS` and `HDRS` variables used before?

tigerw replied:
I, technically, never agreed to #4731 XD

tigerw commented on .gitmodules:
```diff
@@ -47,10 +47,9 @@
-	path = lib/sqlite
-	url = https://github.com/cuberite/sqlite.git
-	ignore = dirty
+[submodule "lib/lsqlite"]
```
Done

tigerw commented on CMake/Fixups.cmake:
```diff
@@ -0,0 +1,14 @@
+# TODO these should be in the submodules
+# Under Windows, we need Lua as DLL; on *nix we need it linked statically:
+if (WIN32)
+	target_compile_definitions(lua PRIVATE LUA_BUILD_AS_DLL)
```
Done

tigerw commented on CMakeLists.txt:
```diff
@@ -13,19 +13,17 @@
+
+option(BUILD_TOOLS "Sets up additional executables to be built along with the server" OFF)
+option(PRECOMPILE_HEADERS "Enable precompiled headers for faster builds" ON)
+option(SELF_TEST "Enables testing code to be built" OFF)
```
Added

tigerw commented on CMakeLists.txt:
```diff
@@ -93,54 +91,20 @@ else()
 endif()
 
+# Static CRT
+set(CMAKE_MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>")
```
It was buried in SetFlags, it's to avoid needing to install a redistributable on users' Windows systems by linking the common runtime statically.

peterbell10 replied:
This seems to require CMake 3.15 or above.

tigerw replied:
Most importantly the AppVeyor images have 3.17, it doesn't really matter for local debugging.

tigerw commented on CMakeLists.txt:
```diff
@@ -156,35 +120,14 @@ if(${BUILD_UNSTABLE_TOOLS})
-include(SetFlags.cmake)
-set_flags()
-set_lib_flags()
-enable_profile()
```
Done

tigerw commented on CMakeLists.txt:
```diff
@@ -156,35 +120,14 @@ if(${BUILD_UNSTABLE_TOOLS})
-endif()
-
-# The Expat library is linked in statically, make the source files aware of that:
-add_definitions(-DXML_STATIC)
```
Moved into submodule https://github.com/cuberite/expat/commit/968efc0b5479b167cec4a7aea8b84ed23adc10c3

madmaxoft replied:
Why was this moved into a submodule, but the other submodules' configurations were left in? (`set(SQLITECPP_RUN_CPPLINT` et al.)

tigerw replied:
It was moved since expat is our repository and there's no real upstream to update from. The others have maintained CMakeLists.texts so I think using their provided configuration options makes more sense.

tigerw commented on CMakeLists.txt:
```diff
@@ -156,35 +120,14 @@ if(${BUILD_UNSTABLE_TOOLS})
-endif()
-
-# Build all dependent libraries as static:
-SET(CMAKE_BUILD_STATIC_LIBRARIES ON)
```
I was added a couple years back but I don't think this variable is a thing :p

There's a BUILD_SHARED_LIBS but I think that's default off anyway; besides we force all of the libraries to build statically via their own `option()`s.

tigerw commented on CMakeLists.txt:
```diff
@@ -156,35 +120,14 @@ if(${BUILD_UNSTABLE_TOOLS})
+# Set the startup project to Cuberite, and the debugger dir:
+set_property(DIRECTORY PROPERTY VS_STARTUP_PROJECT ${CMAKE_PROJECT_NAME})
+set_target_properties(${CMAKE_PROJECT_NAME} PROPERTIES VS_DEBUGGER_WORKING_DIRECTORY "${CMAKE_BINARY_DIR}/Server")
+set_target_properties(${CMAKE_PROJECT_NAME} PROPERTIES UNITY_BUILD ON)
```
Done

tigerw commented on CMakeLists.txt:
```diff
@@ -251,84 +196,66 @@ if (NOT EXISTS ${PROJECT_SOURCE_DIR}/lib/zlib/CMakeLists.txt)
-add_subdirectory(lib/luaexpat/)
-add_subdirectory(lib/libevent/ EXCLUDE_FROM_ALL)
+# Include all the libraries
+# We use EXCLUDE_FROM_ALL so that only the explicit dependencies are used
```
Thanks

tigerw commented on CMakeLists.txt:
```diff
@@ -251,84 +196,66 @@ if (NOT EXISTS ${PROJECT_SOURCE_DIR}/lib/zlib/CMakeLists.txt)
-
-add_subdirectory(src)
-
+# Self Test Mode enables extra checks at startup
 if(${SELF_TEST})
 	message("Tests enabled")
 	enable_testing()
+	target_compile_definitions(${CMAKE_PROJECT_NAME} PRIVATE SELF_TEST)
```
I guess I'll leave it in for now and add a comment?

peterbell10 replied:
I would just get rid of it.

tigerw commented on src/Bindings/AllToLua.pkg:
```diff
@@ -36,11 +36,12 @@ $cfile "../IniFile.h"
-$cfile "Plugin.h"
-$cfile "PluginLua.h"
-$cfile "LuaWindow.h"
+// Extra indirection as CheckBindingsDependencies assumes working directory is sources root:
```
I think it would be best for all the paths in the file to be made relative to src/ but in the interest of small diffs it should be done separately

peterbell10 replied:
We always have includes relative to the current folder. I don't see why we should make an exception here when it was working before. What is breaking exactly?

tigerw replied:
CheckBindings assumes AllToLua and the bindings-generation-CMake-file have paths relative to the same base (src/Bindings), but pulling the latter up into src/ broke the string-based path comparison in one case: "X.cpp" appearing in AllToLua logically means "src/Bindings/X.cpp", and should match a corresponding "Bindings/X.cpp" in the CMake, but the string path comparison only attempts to strip off any `../` and do an equality compare. To fix this I added a redundant "../Bindings" qualifier (the "extra indirection" part) so the strings match.

Making CheckBindings handle paths slightly smarter will work too.

tigerw commented on src/CMakeLists.txt:
```diff
@@ -169,164 +157,45 @@ SET (HDRS
-include_directories("${CMAKE_BINARY_DIR}/include")
-include_directories (SYSTEM "${CMAKE_CURRENT_SOURCE_DIR}/../lib/TCLAP/include")
+# TODO: CMake includes PCH in every source file
+# Remove when every build has PCH enabled:
```
Done

peterbell10 commented on CMake/GenerateBindings.cmake:
```diff
@@ -99,9 +99,9 @@ set(BINDING_OUTPUTS
+set(BINDINGS_FOLDER "${CMAKE_SOURCE_DIR}/src/Bindings/")
 list(TRANSFORM BINDING_OUTPUTS PREPEND ${BINDINGS_FOLDER})
-list(TRANSFORM BINDING_DEPENDENCIES PREPEND "${PROJECT_SOURCE_DIR}/src/")
+list(TRANSFORM BINDING_DEPENDENCIES PREPEND "${CMAKE_SOURCE_DIR}/src/")
```
What happened here?

tigerw replied:
I don't see anything untoward? Sorry if I'm missing something

peterbell10 replied:
Why was it changed from `PROJECT_SOURCE_DIR` to `CMAKE_SOURCE_DIR`. This would disallow including cuberite as a subfolder.

tigerw replied:
exotic
ok I'll change it back

tigerw replied:
dun dun

peterbell10 commented on CMake/Fixups.cmake:
```diff
@@ -0,0 +1,20 @@
+	target_compile_options(mbedcrypto PRIVATE -fomit-frame-pointer)
+endif()
+
+if(CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
+	# Generated file has old-style casts, missing prototypes, and deprecated declarations
+	set_source_files_properties("${CMAKE_SOURCE_DIR}/src/Bindings/Bindings.cpp" PROPERTIES COMPILE_OPTIONS -Wno-everything)
+	set_source_files_properties("${CMAKE_SOURCE_DIR}/src/IniFile.cpp" PROPERTIES COMPILE_OPTIONS -Wno-header-hygiene)
+endif()
```
These aren't fixups for submodules.

tigerw replied:
Indeed, more like fixups for questionable code

tigerw replied:
Would you prefer they went in SetFlags?

peterbell10 replied:
Source file properties should go in the folder's `CMakeLists.txt`.

tigerw replied:
CMake can't propagate them... seriously :(

peterbell10 replied:
Then put it at the top level `CMakeLists.txt` I guess.

tigerw replied:
dun dun dun!

tigerw commented:
lol, clang-tidy broke

peterbell10 commented on CMakeLists.txt:
```diff
@@ -251,84 +193,100 @@ if (NOT EXISTS ${PROJECT_SOURCE_DIR}/lib/zlib/CMakeLists.txt)
-	# https://tls.mbed.org/kb/development/arm-thumb-error-r7-cannot-be-used-in-asm-here
-	target_compile_options(mbedcrypto PRIVATE -fomit-frame-pointer)
+	# File failed to follow NHS guidelines on handwashing and has not maintained good hygiene
+	set_source_files_properties("${CMAKE_SOURCE_DIR}/src/IniFile.cpp" PROPERTIES COMPILE_OPTIONS -Wno-header-hygiene)
```
Wrap in `if (UNITY_BUILDS)`?

tigerw replied:
I think I'll fix the use namespace std, it's cleaner

peterbell10 commented on clang-tidy.sh:
```diff
@@ -8,7 +8,7 @@ ARGS="-header-filter $REGEX -quiet -export-fixes $FIXES_FILE "$@" $REGEX"
 mkdir -p tidy-build
 cd tidy-build
-cmake --target Cuberite -DCMAKE_EXPORT_COMPILE_COMMANDS=ON ..
+cmake --target Cuberite -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DPRECOMPILE_HEADERS=OFF ..
```
Should disable unity builds here I think. It reads the compiled files list to determine which files to lint.

tigerw replied:
Turned off

peterbell10 commented on travisbuild.sh:
```diff
@@ -8,12 +8,18 @@ export CUBERITE_BUILD_DATETIME=`date`
 
 # Use ccache if available
 if [ `which ccache` ]; then
+	# Re-run compile on pre-processed sources on cache miss (slightly faster?)
```
It's slower actually, but clang builds fail without it.

tigerw replied:
I'm starting to think all software sits on a huge creaking tower four miles high and two cm thick.

tigerw replied:
Comment changed

peterbell10 commented on CMake/GenerateBindings.cmake:
```diff
@@ -110,3 +110,11 @@ add_custom_command(
 	DEPENDS ${BINDING_DEPENDENCIES} luaexe
 )
+
+# Generate AllFiles.lst for CheckBasicStyle.lua
+get_target_property(ALL_FILES ${CMAKE_PROJECT_NAME} SOURCES)
+foreach(FILE ${ALL_FILES})
+	file(RELATIVE_PATH RELATIVE "${PROJECT_SOURCE_DIR}/src" ${FILE})
+	set(ALL_FILES_AS_LINES "${ALL_FILES_AS_LINES}${RELATIVE}\n")
+endforeach()
+file(WRITE src/AllFiles.lst "${ALL_FILES_AS_LINES}")
```
This has nothing to do with the bindings so I don't get why you've moved it here.

tigerw replied:
Moved into src/

tigerw commented:
murjhe?

peterbell10 commented on travisbuild.sh:
```diff
@@ -8,12 +8,19 @@ export CUBERITE_BUILD_DATETIME=`date`
-cmake . -DBUILD_TOOLS=1 -DSELF_TEST=1 ${CACHE_ARGS};
+# Work around a Clang + ccache issue with failing builds
+# by disabling precompiled headers
+cmake . -DBUILD_TOOLS=YES -DPRECOMPILE_HEADERS=NO -DSELF_TEST=YES ${CACHE_ARGS};
```
We should also have at least 1 CI run without unity builds.

madmaxoft commented:
Murgh Tandoori?

On a more serious note, how about getting rid of the `{CMAKE_PROJECT_NAME}` and using `Cuberite` directly instead?

tigerw commented:
xoft: the thinking is that it helps with the target-centric commands if I include() a setter file under another project, like the ones in Tools\, that way it will set the properties on the correct target without further modification; also God forbid if we need to change our name again only the project() call needs updating. What do you think?

peterbell: soz, a bunch of comments got hidden automatically and I lost track.

I would present that the version bump is worth it, there is clearly at least one use-case for using 3.13. It's also the default in Debian Buster and Ubuntu Eoan Ermine; Bionic and Xenial can use snaps, and downloading direct from cmake.org/files works fine too.

At this point the libraries are essentially subfolders of the Cuberite folder, our expat for example won't be used by some third party. I think manually specifying minimum versions for them therefore doesn't make sense, we wouldn't put a minimum version in, say, src/OSSupport. I removed them because they were a pessimisation: since the minimums were so old it was preventing CMake from auto-applying CMAKE_INTERPROCEDURAL_OPTIMISATION. Future versions of CMake adding new features may well encounter the same thing, so I think having them be guarded by the top-level Cuberite minimum is ok.

I reverted changes to mbedtls and will do so for libevent, adding in lieu an include_directories workaround, however it is important to note that since the last libevent release a year ago there have been almost 600 commits with bugfixes. From what I recall we've never had issues with libraries being too new (having always updated to master), but libevent networking issues are alongisde threading issues the most difficult to debug since it's hard to reproduce. Then, I think it is worth it trying to get those fixes + performance improvements in. If you insist on releases maybe we could petition the maintainers to cut a newer one?

peterbell10 commented:
>Ubuntu Eoan

For some reason I had this in my mind as 3.12. Even though I'm actually on an eoan-based distro... :facepalm:
In that case, just update the `COMPILING.md`.

>the minimums were so old it was preventing CMake from auto-applying CMAKE_INTERPROCEDURAL_OPTIMISATION.

I would rather you increase the minimum then, rather than removing it completely. IMO the submodules should be as independent as possible, that way we can update the main project CMake without needing to think about the submodules.

>since the last libevent release a year ago there have been almost 600 commits with bugfixes

From their github: 
>This branch is 289 commits ahead, 594 commits behind master. 

They follow a branch based release model where bug-fix patches are cherry-picked from master. I'm sure that any major bugfixes would prompt them to create a new minor release. Either way, I see being hard to debug if things go wrong as a point in favour of stable, tested versions.

tigerw commented on CMakeLists.txt:
```diff
@@ -93,54 +92,20 @@ else()
-set(SELF_TEST OFF CACHE BOOL "")
-
-# Check whether Lua can be used:
-if (NOT(DISABLE_SYSTEM_LUA))
```
CMake resolves the add_custom_command executable automatically to our built one.

tigerw commented:
AppVeyor:

![AppVeyor](https://user-images.githubusercontent.com/2939130/82095683-727b6780-96f7-11ea-92a5-479d280203f2.png)

tigerw commented:
Any final comments from you guys? xoft what do you think about the reasoning for using cmake_project_name?

peterbell10 commented on CMakeLists.txt:
```diff
@@ -251,84 +193,102 @@ if (NOT EXISTS ${PROJECT_SOURCE_DIR}/lib/zlib/CMakeLists.txt)
+add_subdirectory(lib/lua)
+add_subdirectory(lib/luaexpat)
+add_subdirectory(lib/mbedtls)
+add_subdirectory(lib/SQLiteCpp)
```
@tigerw `SQLiteCpp` needs to be included before `lsqlite` so the `sqlite` target is available to link.

tigerw commented:
All good?

peterbell10 commented:
Results from #4742 show cached builds are just as fast and full rebuilds are about 30% faster on travis. Not bad at all.

tigerw commented:
woo! If only precompiled headers worked too 😅

peterbell10 commented:
Can you update `COMPILING.md` and set the submodule commits to the PR I made?

tigerw commented:
I didn't find a reference to CMake 3.12 in COMPILING, it was always 3.13 in there

peterbell10 commented on .travis.yml:
```diff
@@ -22,30 +16,38 @@ jobs:
       compiler: gcc
-      env: *Release
+      before_install:
+        - wget --output-document=${HOME}/CMake http://cmake.org/files/v3.13/cmake-3.13.0-rc1-Linux-x86_64.tar.gz
```
That's a pre-release version.
```suggestion
        - wget --output-document=${HOME}/CMake https://github.com/Kitware/CMake/releases/download/v3.13.0/cmake-3.13.0-Linux-x86_64.tar.gz
```

tigerw commented:
I have attached a yellow "Outdated" label to your most recent comments.
